File: rcc/services/profile_service.py
import time
from threading import Lock


from framework.logger import Logger, logged_method
from framework.singleton import singleton
from rcc.communications.client.cmd.linux.cpupower_client import CPU_POWER_CLIENT
from rcc.communications.client.dbus.asus.core.fan_curves_client import FAN_CURVES_CLIENT
from rcc.communications.client.dbus.asus.core.platform_client import PLATFORM_CLIENT
from rcc.communications.client.cmd.linux.power_profiles_client import POWER_PROFILE_CLIENT
from rcc.communications.client.dbus.linux.upower_client import UPOWER_CLIENT
from rcc.gui.notifier import NOTIFIER
from rcc.models.cpu_governor import CpuGovernor
from rcc.models.fan_curve import FanCurve
from rcc.models.performance_profile import PerformanceProfile
from rcc.models.platform_profile import PlatformProfile
from rcc.models.power_profile import PowerProfile
from rcc.models.ssd_scheduler import SsdScheduler
from rcc.services.hardware_service import HARDWARE_SERVICE
from rcc.utils.beans import EVENT_BUS, TRANSLATOR
from rcc.utils.configuration import CONFIGURATION
from rcc.utils.events import HARDWARE_SERVICE_ON_BATTERY, PLATFORM_SERVICE_PROFILE_CHANGED, STEAM_SERVICE_GAME_EVENT


@singleton
class ProfileService:
    """Service for platform setting"""

    def __init__(self):
        self._logger = Logger()
        self._logger.info("Initializing ProfileService")
        self._logger.add_tab()
        self._lock = Lock()
        self._ac_events_enabled = True

        self._performance_profile = PerformanceProfile(CONFIGURATION.platform.profiles.profile)

        if UPOWER_CLIENT.available:
            self.on_bat = UPOWER_CLIENT.on_battery

        if PLATFORM_CLIENT.available:
            self._platform_profile = PLATFORM_CLIENT.platform_profile
            self._battery_charge_limit = PLATFORM_CLIENT.charge_control_end_threshold

            PLATFORM_CLIENT.change_platform_profile_on_ac = False
            PLATFORM_CLIENT.change_platform_profile_on_battery = False
            PLATFORM_CLIENT.platfom_profile_linked_epp = True

        self.__curves: dict[PlatformProfile, list[FanCurve]] = {}

        if FAN_CURVES_CLIENT.available:
            self._logger.info("Configuring improved fan curves")
            for p in PlatformProfile:
                FAN_CURVES_CLIENT.set_curves_to_defaults(p)
                FAN_CURVES_CLIENT.reset_profile_curves(p)
                time.sleep(0.2)

                curves = FAN_CURVES_CLIENT.fan_curve_data(p)
                for curve in curves:
                    curve.enabled = True
                    for i in range(len(curve.points)):  # pylint: disable=consider-using-enumerate
                        curve.points[i] = (curve.points[i][0], min(100, round(curve.points[i][1] * 1.25)))
                self.__curves[p] = curves

        EVENT_BUS.on(HARDWARE_SERVICE_ON_BATTERY, self._on_ac_battery_change)
        EVENT_BUS.on(STEAM_SERVICE_GAME_EVENT, self._on_game_event)

        self._logger.rem_tab()

    @property
    def performance_profile(self) -> PerformanceProfile:
        """Performance profile"""
        return self._performance_profile

    def _on_game_event(self, count):
        self._ac_events_enabled = count == 0

    def _set_power_profile(self, value):
        self._platform_profile = PowerProfile(value)

    def _on_ac_battery_change(self, on_battery: bool, force=False):
        if self._ac_events_enabled or force:
            self.on_bat = on_battery
            self.set_performance_profile(PerformanceProfile.QUIET, True, True)

    @logged_method
    def set_performance_profile(  # pylint: disable=too-many-locals
        self, profile: PerformanceProfile, temporal=False, force=False
    ) -> None:
        """Establish performance profile"""
        with self._lock:
            profile_name = profile.name
            if self._performance_profile != profile or force:
                platform_profile = profile.platform_profile
                power_profile = profile.power_profile
                cpu_governor = profile.battery_governor if self.on_bat else profile.ac_governor
                boost_enabled = profile.battery_boost if self.on_bat else profile.ac_boost
                ssd_sched = profile.ssd_queue_scheduler
                try:
                    self._logger.info(f"Setting {profile_name} profile")  # pylint: disable=line-too-long
                    t0 = time.time()
                    self._logger.add_tab()

                    self.__set_throttle_policy(profile, platform_profile)
                    self.__set_tdp_values(profile)
                    self.__set_fan_curves(platform_profile)
                    self.__set_boost(boost_enabled)
                    self.__set_cpu_governor(cpu_governor, power_profile)
                    self.__set_ssd_scheduler(ssd_sched)

                    self._logger.rem_tab()
                    self._logger.info(f"Profile setted succesfully after {round(1000*(time.time()-t0))/1000} ms")

                    if not temporal and not self.on_bat:
                        CONFIGURATION.platform.profiles.profile = profile.value
                        CONFIGURATION.save_config()

                    NOTIFIER.show_toast(
                        TRANSLATOR.translate(
                            "profile.applied",
                            {"profile": TRANSLATOR.translate(f"label.profile.{profile_name}").lower()},
                        )
                    )

                    self._performance_profile = profile
                    EVENT_BUS.emit(
                        PLATFORM_SERVICE_PROFILE_CHANGED,
                        self._performance_profile,
                    )
                except Exception as error:
                    self._logger.error(f"Couldn't set profile: {error}")
                    self._logger.rem_tab()
            else:
                self._logger.info(f"Profile {profile_name.lower()} already setted")

    def restore_profile(self):
        """Restore persited profile"""
        if self.on_bat:
            self._logger.info("Laptop running on battery")
            self._logger.add_tab()
            self._on_ac_battery_change(True, True)
            self._logger.rem_tab()
        else:
            self._logger.info("Laptop running on AC")
            self._logger.add_tab()
            self.set_performance_profile(PerformanceProfile(CONFIGURATION.platform.profiles.profile), True, True)
            self._logger.rem_tab()

    def __set_throttle_policy(self, profile: PerformanceProfile, platform_profile: PlatformProfile):
        try:
            platform_profile = profile.platform_profile
            self._logger.info(f"Throttle policy: {platform_profile.name}")
            self._logger.add_tab()

            if PLATFORM_CLIENT.available:
                self._platform_profile = platform_profile
                PLATFORM_CLIENT.platform_profile = platform_profile
                PLATFORM_CLIENT.enable_ppt_group = True

            self._logger.rem_tab()
        except Exception as e:
            self._logger.error(f"Error while setting platform policy: {e}")

    def __set_tdp_values(self, profile: PerformanceProfile):
        if HARDWARE_SERVICE.tdp_available:
            try:
                self._logger.info("TDP values")
                self._logger.add_tab()
                HARDWARE_SERVICE.set_cpu_tdp(profile)
                HARDWARE_SERVICE.set_gpu_tgp(profile)
                self._logger.rem_tab()
            except Exception as e:
                self._logger.error(f"Error while setting TDP values: {e}")

    def __set_fan_curves(self, platform_profile: PlatformProfile):
        if FAN_CURVES_CLIENT.available:
            try:
                self._logger.info(f"Fan curve: {platform_profile.name}")
                for curve in self.__curves[platform_profile]:
                    FAN_CURVES_CLIENT.set_fan_curve(platform_profile, curve)
            except Exception as e:
                self._logger.error(f"Error while setting fan curve: {e}")

    def __set_cpu_governor(self, governor: CpuGovernor, power_profile: PowerProfile):
        if CPU_POWER_CLIENT.available:
            try:
                self._logger.info(f"CPU governor: {governor.name}")
                CPU_POWER_CLIENT.set_governor(governor)
                time.sleep(0.25)
                if POWER_PROFILE_CLIENT.available:
                    self._logger.info(f"Power profile: {power_profile.name}")
                    POWER_PROFILE_CLIENT.active_profile = power_profile
            except Exception as e:
                self._logger.error(f"Error while setting CPU governor: {e}")

    def __set_boost(self, boost_enabled: bool):
        try:
            HARDWARE_SERVICE.set_boost_status(boost_enabled)
        except Exception as e:
            self._logger.error(f"Error while setting boost mode: {e}")

    def __set_ssd_scheduler(self, scheduler: SsdScheduler):
        try:
            HARDWARE_SERVICE.set_ssd_scheduler(scheduler)
        except Exception as e:
            self._logger.error(f"Error while setting SSD scheduler: {e}")


PROFILE_SERVICE = ProfileService()

File: core/src/services/profile_service.cpp

#include "../../include/clients/dbus/asus/core/fan_curves_client.hpp"
#include "../../include/clients/dbus/asus/core/platform_client.hpp"
#include "../../include/clients/dbus/linux/upower_client.hpp"
#include "../../include/clients/file/boost_control_client.hpp"
#include "../../include/clients/file/ssd_scheduler_client.hpp"
#include "../../include/clients/shell/cpupower_client.hpp"
#include "../../include/clients/shell/power_profiles_client.hpp"
#include "../../include/gui/toaster.hpp"
#include "../../include/services/hardware_service.hpp"
#include "../../include/services/profile_service.hpp"
#include "../../include/utils/profile_utils.hpp"
#include "RccCommons.hpp"

ProfileService::ProfileService()
{
    logger.info("Initializing ProfileService");
    logger.add_tab();

    currentProfile = Configuration::getInstance().getConfiguration().platform.profiles.profile;

    if (UPowerClient::getInstance().available())
    {
        onBattery = UPowerClient::getInstance().isOnBattery();
        std::string mode = onBattery ? "battery" : "AC";
        logger.info("Laptop on " + mode + " mode");
    }

    if (PlatformClient::getInstance().available())
    {
        PlatformClient::getInstance().setChangePlatformProfileOnAc(false);
        PlatformClient::getInstance().setChangePlatformProfileOnBattery(false);
        PlatformClient::getInstance().setPlatformProfileLinkedEpp(true);
    }

    logger.rem_tab();
}

PerformanceProfile ProfileService::getPerformanceProfile()
{
    return currentProfile;
}
void ProfileService::setPerformanceProfile(PerformanceProfile profile, bool temporal, bool force)
{
    std::lock_guard<std::mutex> lock(mutex);
    std::string profileName = profile.toName();

    if (profile != currentProfile || force)
    {
        logger.info("Setting " + profileName + " profile");
        logger.add_tab();
        try
        {
            auto t0 = std::chrono::high_resolution_clock::now();

            setPlatformProfile(profile);
            // setFanCurves(profile);
            setBoost(profile);
            setCpuGovernor(profile);
            setSsdScheduler(profile);
            setPowerProfile(profile);
            setTdps(profile);
            setTgp(profile);

            currentProfile = profile;
            Configuration::getInstance().getConfiguration().platform.profiles.profile = profile;
            Configuration::getInstance().saveConfig();

            auto t1 = std::chrono::high_resolution_clock::now();
            logger.rem_tab();
            logger.info("Profile setted after " + std::to_string(std::chrono::duration_cast<std::chrono::milliseconds>(t1 - t0).count()) + " ms");
            std::unordered_map<std::string, std::any> values = {{"profile", StringUtils::toLowerCase(Translator::getInstance().translate("label.profile." + profileName))}};
            Toaster::getInstance().showToast(Translator::getInstance().translate("profile.applied", values));
        }
        catch (std::exception e)
        {
            logger.rem_tab();
        }
    }
    else
    {
        logger.info("Profile " + StringUtils::toLowerCase(profileName) + " already setted");
    }
}

void ProfileService::setPlatformProfile(PerformanceProfile profile)
{
    if (PlatformClient::getInstance().available())
    {
        auto platformProfile = ProfileUtils::platformProfile(profile);
        logger.info("Platform profile: {}", platformProfile.toName());
        logger.add_tab();
        try
        {
            PlatformClient::getInstance().setPlatformProfile(platformProfile);
            PlatformClient::getInstance().setEnablePptGroup(true);
        }
        catch (std::exception e)
        {
            logger.error("Error while platform profile: {}", e.what());
        }
        logger.rem_tab();
    }
}

void ProfileService::setFanCurves(PerformanceProfile profile)
{
    if (FanCurvesClient::getInstance().available())
    {
        auto platformProfile = ProfileUtils::platformProfile(profile);
        logger.info("Fan profile: {}", platformProfile.toName());
        logger.add_tab();
        try
        {
            FanCurvesClient::getInstance().resetProfileCurve(platformProfile);
            FanCurvesClient::getInstance().setCurveToDefaults(platformProfile);
            FanCurvesClient::getInstance().setFanCurveEnabled(platformProfile);
        }
        catch (std::exception e)
        {
            logger.error("Error while setting fan curve: {}", std::string(e.what()));
        }
        logger.rem_tab();
    }
}

void ProfileService::setBoost(PerformanceProfile profile)
{
    if (BoostControlClient::getInstance().available())
    {
        bool enabled = onBattery ? ProfileUtils::batteryBoost() : ProfileUtils::acBoost();
        logger.info("CPU boost: {}", enabled ? "ON" : "OFF");
        logger.add_tab();
        try
        {
            BoostControlClient::getInstance().set_boost(enabled);
        }
        catch (std::exception e)
        {
            logger.error("Error while setting CPU boost: {}", std::string(e.what()));
        }
        logger.rem_tab();
    }
}

void ProfileService::setSsdScheduler(PerformanceProfile profile)
{
    if (SsdSchedulerClient::getInstance().available())
    {
        SsdScheduler ssdScheduler = ProfileUtils::ssdQueueScheduler(profile);
        logger.info("SSD scheduler: {}", ssdScheduler.toName());
        logger.add_tab();
        try
        {
            SsdSchedulerClient::getInstance().setScheduler(ssdScheduler);
        }
        catch (std::exception e)
        {
            logger.error("Error while setting SSD scheduler: {}", std::string(e.what()));
        }
        logger.rem_tab();
    }
}

void ProfileService::setCpuGovernor(PerformanceProfile profile)
{
    if (CpuPowerClient::getInstance().available())
    {
        CpuGovernor cpuGovernor = onBattery ? ProfileUtils::batteryGovernor() : ProfileUtils::acGovernor(profile);
        logger.info("CPU governor: {}", cpuGovernor.toName());
        logger.add_tab();
        try
        {
            CpuPowerClient::getInstance().setGovernor(cpuGovernor);
        }
        catch (std::exception e)
        {
            logger.error("Error while setting CPU governor: {}", std::string(e.what()));
        }
        logger.rem_tab();
    }
}

void ProfileService::setPowerProfile(PerformanceProfile profile)
{
    if (PowerProfileClient::getInstance().available())
    {
        PowerProfile powerProfile = ProfileUtils::powerProfile(profile);
        logger.info("Power profile: {}", powerProfile.toName());
        logger.add_tab();
        try
        {
            PowerProfileClient::getInstance().setProfile(powerProfile);
        }
        catch (std::exception e)
        {
            logger.error("Error while setting power profile: {}", std::string(e.what()));
        }
        logger.rem_tab();
    }
}

void ProfileService::setTdps(PerformanceProfile profile)
{
    if (Pl1SpdClient::getInstance().available())
    {
        logger.info("TDP values");
        logger.add_tab();

        auto pl1 = onBattery ? ProfileUtils::batteryIntelPl1Spl(profile) : ProfileUtils::acIntelPl1Spl(profile);
        logger.info("PL1: " + std::to_string(pl1) + "W");
        try
        {
            std::this_thread::sleep_for(std::chrono::milliseconds(50));
            Pl1SpdClient::getInstance().setCurrentValue(pl1);
            if (Pl2SpptClient::getInstance().available())
            {
                auto pl2 = onBattery ? ProfileUtils::batteryIntelPl2Sppt(profile) : ProfileUtils::acIntelPl2Sppt(profile);
                logger.info("PL2: " + std::to_string(pl2) + "W");
                std::this_thread::sleep_for(std::chrono::milliseconds(50));
                Pl2SpptClient::getInstance().setCurrentValue(pl2);
                std::this_thread::sleep_for(std::chrono::milliseconds(50));
            }
        }
        catch (std::exception e)
        {
            logger.info("Error setting CPU TDPs");
        }

        logger.rem_tab();
    }
}

void ProfileService::setTgp(PerformanceProfile profile)
{
    if (NvTempClient::getInstance().available() || NvBoostClient::getInstance().available())
    {
        logger.info("Nvidia GPU");
        logger.add_tab();

        if (NvBoostClient::getInstance().available())
        {
            try
            {
                auto nvb = onBattery ? ProfileUtils::batteryNvBoost(profile) : ProfileUtils::acNvBoost(profile);
                logger.info("Dynamic Boost: " + std::to_string(nvb) + "ºC");
                NvBoostClient::getInstance().setCurrentValue(nvb);
            }
            catch (std::exception e)
            {
                logger.info("Error setting Nvidia Boost");
            }
        }

        if (NvTempClient::getInstance().available())
        {
            try
            {
                auto nvt = onBattery ? ProfileUtils::batteryNvTemp(profile) : ProfileUtils::acNvTemp();
                logger.info("Throttle temp: " + std::to_string(nvt) + "ºC");
                NvTempClient::getInstance().setCurrentValue(nvt);
            }
            catch (std::exception e)
            {
                logger.info("Error setting Nvidia TGP");
            }
        }

        logger.rem_tab();
    }
}
